src/Jaeger.Core/Tracer.cs

```diff
@@ -18,189 +20,292 @@
+
+        public string Version { get; }
+        public bool ZipkinSharedRpcSpan { get; }
+        public bool ExpandExceptionLogs { get; }
+        public long IPv4 { get; }
+
+        private Tracer(
+            string serviceName,
+            IReporter reporter,
+            ISampler sampler,
+            PropagationRegistry registry,
+            IClock clock,
+            IMetrics metrics,
+            ILoggerFactory loggerFactory,
+            Dictionary<string, object> tags,
+            bool zipkinSharedRpcSpan,
+            IScopeManager scopeManager,
+            IBaggageRestrictionManager baggageRestrictionManager,
+            bool expandExceptionLogs)
         {
             ServiceName = serviceName;
-            Tags = tags;
-            ScopeManager = scopeManager;
-            PropagationRegistry = propagationRegistry;
-            Sampler = sampler;
             Reporter = reporter;
+            Sampler = sampler;
+            Registry = registry;
+            Clock = clock;
             Metrics = metrics;
-            Clock = new Clock();
+            Logger = loggerFactory.CreateLogger<Tracer>();
+            ZipkinSharedRpcSpan = zipkinSharedRpcSpan;
+            ScopeManager = scopeManager;
+            _baggageSetter = new BaggageSetter(baggageRestrictionManager, metrics);
+            ExpandExceptionLogs = expandExceptionLogs;
 
-            _logger = loggerFactory.CreateLogger<Tracer>();
+            Version = LoadVersion();
+            tags[Constants.JaegerClientVersionTagKey] = Version;
 
-            if (tags.TryGetValue(Constants.TracerIpTagKey, out object field))
+            string hostname = GetHostName();
+            if (!tags.ContainsKey(Constants.TracerHostnameTagKey))
             {
-                HostIPv4 = field.ToString();
+                if (hostname != null)
+                {
+                    tags[Constants.TracerHostnameTagKey] = hostname;
+                }
             }
-        }
 
-        private static string GetVersion()
-        {
-            var version = Assembly.GetExecutingAssembly().GetName().Version.ToString();
-            return $"CSharp-{version}";
+            if (tags.TryGetValue(Constants.TracerIpTagKey, out object ipTag))
+            {
+                try
+                {
+                    IPv4 = Utils.IpToInt(ipTag as string);
+                }
+                catch
+                {
+                }
+            }
+            else
+            {
+                try
+                {
+                    IPAddress hostIPv4 = Dns.GetHostAddresses(hostname).First(ip => ip.AddressFamily == AddressFamily.InterNetwork);
+
+                    tags[Constants.TracerIpTagKey] = hostIPv4.ToString();
+                    IPv4 = Utils.IpToInt(hostIPv4);
+                }
+                catch
+                {
+                }
+            }
+
+            Tags = new ReadOnlyDictionary<string, object>(tags);
```

Comment:
`Dictionary` already implements `IReadOnlyDictionary`. Wrapping in `ReadOnlyDictionary` not needed.